### PR TITLE
command: Don't update backend hash unless backend is changing

### PR DESCRIPTION
**Pull request by chrisarcand**
The `Meta.backend_C_r_S_unchanged()` method is sadly a bit of a mess.

It seems to have originally been used as a method to be called when the backend is not changing, with an extra assumption that if the configured backend's hash doesn't match the one in state, surely the hash should just be updated as an option might have been moved to command line flags.

However, this function is used throughout this file as 'the method to load the initialized (but not necessarily configured) backend', regardless of whether or not it is the same (unchanged). This is in addition to `Meta.backendFromState()`, which is used to load the same thing except in the main codepath of 'init -backend=false'.

A more proper way to handle this situation would be to consolidate this method (and `backendFromState()`) in to a shared API that pulls these responsibilities apart. One for loading the backend from state, one for updating the hash, etc etc. In the interests of stability and deferring larger refactoring, however, we'll cheat here to maintain the same odd API responsibilities of this method and just not update the hash if the backends don't match, which they frequently don't.

This is awkward, as the API here really assumes the caller wouldn't be calling this method unless that was already the case (it is literally named 'unchanged') but is a reasonable workaround to avoid changing this entire interface to separate that out.

### Testing plan

Basically, any error in the middle of a state migration should _not_ cause the hash value in the legacy state file to change. That is, any error should abort the migration process and you should be able to restart the process by re-running `init`. Previously, re-running `init` would result in a successful init of the old backend with a new configuration, because the hash value would reflect the new one while the backend name remains unchanged.

An easy way to test this is to attempt a state migration (using `tags`) with Terraform Cloud, and give an intentionally bad value for one of the prompts. Re-running init should restart the process, not succeed with the local backend.

### CODE REVIEW

**Comment by jbardin** [comment]
I think this change works for the case presented, but if I understand the situation correctly, I don't think there is a general way to solve the problem since migration can still be happening between backends of the same `Type`. We need to carefully consider what sort of migration behavior we will want when it comes time to refactor out the backend code, as I have a feeling that the existing automatic migration scheme has caused just as many problems as it has solved (`-migrate-state` was a first step in that direction, but it unfortunately still had to use the old code to do it).

**Review by barrettclark — APPROVED** [review]
This looks good. I tried a few different paths that caused the error state, and they all look good now. However, there may be an quirky issue with migrating multiple workspaces to TFC. I'll mention that in Slack.

**Comment by alisdair** [comment]
I agree with James that there doesn't seem to be a way of solving the entire problem without a significant restructure. For example, this approach doesn't solve the other reported issues with the backend hash being updated too early (#26259, #23093).

I'd be fine with merging this if it fixes the problem at hand without making any others worse, but I can't be sure from reading the code that this is the case. Maybe enumerating which situations you've tested with this branch would help to clarify how safe a change it is.

**Comment by chrisarcand** [comment]
I went a different direction with this that handles the bugs Alisdair linked above in a broader way, so closing this in favor of #29860
